quinn/functions.py

```diff
@@ -95,3 +95,248 @@ def array_choice(col):
+            keys,
+        )
+    )
+    drop_column_list = [col_name]
+    df = df.select(
+        [
+            col_to_keep for col_to_keep in df.columns if col_to_keep not in drop_column_list
+        ]
+        + key_cols
+    )
+    return df
+
+def flatten_dataframe(
+    df: DataFrame,
+    sep: str = ":",
+    sort_columns: bool = True,
+    replace_char: str = "_"
+) -> DataFrame:
+    """
+    Flattens all complex data types (StructType, ArrayType, and MapType) in the input DataFrame and returns a
+    new DataFrame with the flattened columns.
+    
+    Args:
+        df (DataFrame): The input PySpark DataFrame.
+        sep (str, optional): The separator to use in the resulting flattened column names. Defaults to ':'.
+        sort_columns (bool, optional): Sort the columns alphabetically. Defaults to False.
+        replace_char (str, optional): The character to replace special characters with in column names. Defaults to '_'.
+        
+    Returns:
+        DataFrame: The DataFrame with all complex data types flattened.
+
+    Note:
+        This function assumes the input DataFrame has a consistent schema across all rows. If you have files with
+        different schemas, use the read_and_flatten_nested_files function instead.
+
+    Example:
+    >>> data = [
+            (
+                1,
+                ("Alice", 25),
+                {"A": 100, "B": 200},
+                ["apple", "banana"],
+                {"key": {"nested_key": 10}},
+                {"A#": 1000, "B@": 2000},
+            ),
+            (
+                2,
+                ("Bob", 30),
+                {"A": 150, "B": 250},
+                ["orange", "grape"],
+                {"key": {"nested_key": 20}},
+                {"A#": 1500, "B@": 2500},
+            ),
+        ]
+    >>> schema = StructType([
+            StructField("id", IntegerType(), True),
+            StructField("name_age", StructType([
+                StructField("name", StringType(), True),
+                StructField("age", IntegerType(), True)
+            ]), True),
+            StructField("scores", MapType(StringType(), IntegerType()), True),
+            StructField("fruits", ArrayType(StringType()), True),
+            StructField("nested_key_map", MapType(StringType(), MapType(StringType(), IntegerType())), True),
+            StructField("special_chars_map", MapType(StringType(), IntegerType()), True),
+        ])
+    >>> df = spark.createDataFrame(data, schema)
+    >>> flattened_df = flatten_dataframe(df)
+    >>> flattened_df.show()
+    >>> flattened_df_with_hyphen = flatten_dataframe(df, replace_char="-")
+    >>> flattened_df_with_hyphen.show()
+    """
+    complex_fields = get_complex_fields(df)
+
+    while len(complex_fields) != 0:
+        col_name = list(complex_fields.keys())[0]
+
+        if isinstance(complex_fields[col_name], StructType):
+            df = flatten_structs(df, col_name, complex_fields, sep, replace_char)
```

Comment:
May we rewrite it in a way when we are not modifying `df` in a loop but instead we fill the list of column and do only one `select` in the end? Cause a lot of modifications can tend to `StackOverflow` errors in Catalyst and also may slow down performance...


Like instead of writing `df = flatten_strcuts(...)` we are using `list_to_select.extend(flatten_structs(...))` and only single select operation (`return df.select(*list_to_select)`).


I know that it requires a lot of changes but I really afraid that using modification of `df` in a loop may cause `StackOverfFlow` or at least tend to veeeeeery long step of pruning logical plan.

Reply:
ok, thank you. I will have a look at this issue.